### PR TITLE
(proof-new) Proof node updater

### PR DESCRIPTION
**Pull request by ajreynol**
Adds utility for updating proof nodes. The module for post-processing proof nodes in the SmtEngine for the sake of proof conversion to external formats will build on this utility.  

Requires https://github.com/CVC4/CVC4/pull/4617.

### CODE REVIEW

File: src/expr/proof_node_updater.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*********************                                                        */
+/*! \file proof_node_updater.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
```
Comment (HanielB) [inline src/expr/proof_node_updater.h:7]:
```suggestion
 ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
```

File: src/expr/proof_node_updater.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  /** post-process */
+  void process(std::shared_ptr<ProofNode> pf);
+
+ private:
+  /** The proof node manager */
+  ProofNodeManager* d_pnm;
+  /** The callback */
+  ProofNodeUpdaterCallback& d_cb;
+  /** Kinds of proof rules we are eliminating */
+  // std::unordered_set<PfRule, PfRuleHashFunction> d_elimRules;
```
Comment (HanielB) [inline src/expr/proof_node_updater.h:70]: delete?

File: src/expr/proof_node_updater.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of a utility for updating proof nodes
+ **/
+
+#include "expr/proof_node_updater.h"
+
+#include "expr/lazy_proof.h"
+
+using namespace CVC4::kind;
```
Comment (HanielB) [inline src/expr/proof_node_updater.cpp:19]:
```suggestion
```

File: src/expr/proof_node_updater.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    cur = visit.back();
+    it = visited.find(cur);
+    if (it == visited.end())
+    {
+      visited.insert(cur);
+      // should it be updated?
+      if (d_cb.shouldUpdate(cur))
+      {
+        PfRule id = cur->getRule();
+        LazyCDProof lcp(d_pnm);
```
Comment (HanielB) [inline src/expr/proof_node_updater.cpp:51]: say why lazy proof

File: src/expr/proof_node_updater.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ * class. It runs this callback on all subproofs of a provided ProofNode
+ * application that meet some criteria (ProofNodeUpdaterCallback::shouldUpdate)
+ * and overwrites them based on the update procedure of the callback
+ * (ProofNodeUpdaterCallback::update).
+ */
+class ProofNodeUpdater
+{
+ public:
+  ProofNodeUpdater(ProofNodeManager* pnm, ProofNodeUpdaterCallback& cb);
+  /** post-process */
```
Comment (HanielB) [inline src/expr/proof_node_updater.h:61]: Detail how this work in terms of the data-structures it uses.

File: src/expr/proof_node_updater.cpp (hunk 4 above)
Comment (ajreynol) [reply src/expr/proof_node_updater.cpp]: Didn't have to be lazy proof, I've updated.

**Comment by HanielB** [comment]
BTW you should update the year in the cpp too.

**Pull request merged**